File: src/com/haskforce/HaskellModuleType.java
package com.haskforce;

import com.intellij.ide.util.projectWizard.ModuleWizardStep;
import com.intellij.ide.util.projectWizard.ProjectJdkForModuleStep;
import com.intellij.ide.util.projectWizard.WizardContext;
import com.intellij.openapi.module.ModuleType;
import com.intellij.openapi.module.ModuleTypeManager;
import com.intellij.openapi.roots.ui.configuration.ModulesProvider;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

public class HaskellModuleType extends ModuleType<HaskellModuleBuilder> {
    public static final String MODULE_TYPE_ID = "HASKELL_MODULE";

    public HaskellModuleType() {
        super(MODULE_TYPE_ID);
    }

    public static HaskellModuleType getInstance() {
        return (HaskellModuleType) ModuleTypeManager.getInstance().findByID(MODULE_TYPE_ID);
    }

    @NotNull
    @Override
    public HaskellModuleBuilder createModuleBuilder() {
        return new HaskellModuleBuilder();
    }

    @NotNull
    @Override
    public String getName() {
        return "Haskell Module";
    }

    @NotNull
    @Override
    public String getDescription() {
        return "Haskell modules are used for developing <b>Haskell</b> applications.";
    }

    @Override
    public Icon getBigIcon() {
        return HaskellIcons.FILE;
    }

    @Override
    public Icon getNodeIcon(@Deprecated boolean isOpened) {
        return HaskellIcons.FILE;
    }
}

File: src/com/haskforce/haskell/HaskellModuleType.scala
package com.haskforce.haskell

import javax.swing.Icon

import com.haskforce.haskell.ui.HaskellIcons
import com.intellij.openapi.module._
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.{VfsUtilCore, VirtualFile}

import scala.collection.JavaConverters._

object HaskellModuleType {
  val MODULE_TYPE_ID = "HASKELL_MODULE"

  def getInstance: HaskellModuleType = {
    ModuleTypeManager.getInstance().findByID(MODULE_TYPE_ID).asInstanceOf[HaskellModuleType]
  }

  def findModules(project: Project): java.util.Collection[Module] = {
    ModuleUtil.getModulesOfType(project, HaskellModuleType.getInstance)
  }

  //TODO review ModuleFile
  def findModule(location: VirtualFile, project: Project): Option[Module] = {
    ModuleUtil.getModulesOfType(project, HaskellModuleType.getInstance)
      .asScala.filter(module => VfsUtilCore.isAncestor(module.getModuleFile.getParent, location, false))
      .reduceOption((module1, module2) => {
        VfsUtilCore.isAncestor(module1.getModuleFile.getParent, module2.getModuleFile.getParent, true) match {
          case true => module2
          case false => module1
        }
      })
  }
}

class HaskellModuleType extends ModuleType[HaskellModuleBuilder](HaskellModuleType.MODULE_TYPE_ID) {
    def createModuleBuilder(): HaskellModuleBuilder = new HaskellModuleBuilder()
    def getName: String = "Haskell Module"
    def getDescription: String = "Haskell modules are used for developing <b>Haskell</b> applications."
    def getBigIcon: Icon = HaskellIcons.FILE
    def getNodeIcon(@Deprecated isOpened: Boolean) = HaskellIcons.FILE
}
